### PR TITLE
Introduce installations

### PR DESCRIPTION
**Pull request by jferris**
Because:

* Multiple users belong to the same Namely account
* Only one connection can sync for each integration/Namely account

This commit:

* Introduces an invisible installation concept
* Adds users from the same Namely account to the same installation
* Associates connections with installations instead of users

Effects:

* Multiple users on the same Namely account can share connections

https://trello.com/c/W6KqJcez

### CODE REVIEW

File: app/jobs/sync_job.rb [hunk 1]
```diff
@@ -14,26 +14,28 @@ def perform
   private
 
   def connection
-    user.send(:"#{@integration_id}_connection")
+    installation.send(:"#{@integration_id}_connection")
```
Comment (sikachu) [inline app/jobs/sync_job.rb]: Do you think it's necessary to use `#send`? Or, will it work with `#public_send`?

Comment (jferris) [inline app/jobs/sync_job.rb]: It works with `public_send`. Addressed in 843cec9.

Comment (kenyonj) [inline app/jobs/sync_job.rb]: I don't think you need the preceeding `:` there... Does that do anything different than `.send("#{@integration_id}_connection")` ?

File: app/jobs/sync_job.rb [hunk 2]
```diff
@@ -14,26 +14,28 @@ def perform
   private
 
   def connection
-    user.send(:"#{@integration_id}_connection")
+    installation.public_send(:"#{@integration_id}_connection")
```
Comment (kenyonj) [inline app/jobs/sync_job.rb]: I don't think you need the preceeding `:` there. Does that do anything different than `public_send("#{@integration_id}_connection")` ?

File: app/models/installation.rb [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+  validates :subdomain, presence: true, uniqueness: true
+
+  def self.ready_to_sync_with(integration)
+    association = "#{integration}_connection"
+    joins(association.to_sym).
+      where(association.pluralize => { found_namely_field: true })
+  end
+
+  def jobvite_connection
+    super || Jobvite::Connection.create(installation: self)
```
Comment (sikachu) [inline app/models/installation.rb]: Do you think this would be better as `create!`? I feel like you could ended up with an unpersisted record here.

File: app/jobs/sync_job.rb [hunk 4]
```diff
@@ -14,26 +14,28 @@ def perform
...
       integration_id: integration_id,
-      user: user
+      installation: installation
     )
   end
 
-  def user
-    @user ||= User.find(@user_id)
+  def installation
+    @installation ||= Installation.find(@installation_id)
```
Comment (kenyonj) [inline app/jobs/sync_job.rb]: thoughts on moving this over to `@_installation` to designate that you won't be using it directly?

File: app/jobs/sync_job.rb (hunk 2 above)
Comment (jferris) [inline app/jobs/sync_job.rb]: It will work with or without the `:`. Using the `:` will generate a symbol, whereas not using it will generate a string.

In general, using symbols for a closed set of literals is more efficient, because additional strings won't be constructed and comparison can be performed by identity (because symbols are immutable).

In this case, we're dynamically creating a symbol (although the set is still closed). I'm unsure of the performance characteristics for dynamic internalization.

Either way, I don't believe this code path is performance-critical enough that we need to worry about string allocation. Because building a string is slightly simpler, I removed the symbol in 6066095.

File: app/jobs/sync_job.rb (hunk 4 above)
Comment (jferris) [inline app/jobs/sync_job.rb]: We haven't been following that convention in this project, so I don't believe code authors will understand that designation.

File: app/models/installation.rb (hunk 3 above)
Comment (jferris) [inline app/models/installation.rb]: Yes, I think `create!` is likely better, since we don't handle the failure case.

This code was extracted from `User`, but I'll see if it breaks anything to switch to `create!`.

File: db/migrate/20150807150624_adjust_user_add_reference_to_installation.rb [hunk 5]
```diff
@@ -0,0 +1,19 @@
+class AdjustUserAddReferenceToInstallation < ActiveRecord::Migration
+  def up
+    add_reference :users, :installation, index: true, foreign_key: true
+
+    update(<<-SQL)
+      UPDATE users SET installation_id = (
```
Comment (sikachu) [inline db/migrate/20150807150624_adjust_user_add_reference_to_installation.rb]: How many users are we going to run this migration against? This subquery could take quite some time to run.

File: app/models/installation.rb (hunk 3 above)
Comment (jferris) [inline app/models/installation.rb]: The tests pass and the UI continues to respond using `create!`, so I introduced it in 035782c.

File: db/migrate/20150807150624_adjust_user_add_reference_to_installation.rb (hunk 5 above)
Comment (jferris) [inline db/migrate/20150807150624_adjust_user_add_reference_to_installation.rb]: The number of users is fairly low, and there are indexes available for the search part of this query.

File: spec/factories.rb [hunk 6]
```diff
@@ -75,5 +90,10 @@
     sequence(:refresh_token) { |n| "refresh-token-#{n}" }
     access_token_expiry { 15.minutes.from_now }
     email "integrationlover@example.com"
+    installation { Installation.find_or_create_by!(subdomain: subdomain) }
```
Comment (sikachu) [inline spec/factories.rb]: Wouldn't this make it create a new subdomain record even if you do `build_stubbed`?

File: app/jobs/sync_job.rb [hunk 7]
```diff
@@ -14,26 +14,28 @@ def perform
   private
 
   def connection
-    user.send(:"#{@integration_id}_connection")
+    installation.public_send("#{@integration_id}_connection")
```
Comment (seanpdoyle) [inline app/jobs/sync_job.rb]: Would a private `attr_reader` make sense for `@integration_id`?

File: app/services/candidate_importer.rb [hunk 8]
```diff
@@ -32,38 +32,42 @@ def import_assistant
...
       integration_id: assistant_class::INTEGRATION_ID,
-      user: user
+      installation: installation
     )
   end
 
-  def user
-    connection.user
+  def installation
+    connection.installation
```
Comment (kenyonj) [inline app/services/candidate_importer.rb]: It looks like `installation` is being used multiple times, should this be memoized?

File: spec/models/installation_spec.rb [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+        :installation,
+        subdomain: "ready-with-given-service"
+      )
+      create(
+        :net_suite_connection,
+        :connected,
+        :with_namely_field,
+        installation: installation
+      )
+      installation = create(:installation, subdomain: "connected-without-field")
```
Comment (sikachu) [inline spec/models/installation_spec.rb]: These setups are a bit hard to follow. I feel like it'd be better put in a well named method instead, but I'll let you decide.

File: app/jobs/sync_job.rb (hunk 7 above)
Comment (seanpdoyle) [inline app/jobs/sync_job.rb]: Looks like that's already the case on line `41`.

Could we use that here?

File: spec/models/authentication_notifier_spec.rb [hunk 10]
```diff
@@ -16,28 +16,29 @@
...
-        user: user_spy
+        installation: installation_spy
       )
       exception = Unauthorized.new(Unauthorized::DEFAULT_MESSAGE)
 
       expect(Rails.logger).to receive(:error).with(
         "#{exception.class} error #{exception.message} for " \
-        "user_id: #{user_spy.id} with #{notifier.integration_name}"
+        "installation_id: #{installation_spy.id} " \
+        "with #{notifier.integration_name}"
```
Comment (seanpdoyle) [inline spec/models/authentication_notifier_spec.rb]: Would it make sense to match against a regex, in case minor changes are made to the error message?

File: spec/models/installation_spec.rb [hunk 11]
```diff
@@ -0,0 +1,157 @@
+require "rails_helper"
+
+describe Installation do
+  describe "validations" do
+    subject { create(:installation) }
+    it { should validate_presence_of(:subdomain) }
```
Comment (seanpdoyle) [inline spec/models/installation_spec.rb]: We've used `is_expected.to` elsewhere in the codebase.

Would it make sense to standardize to one particular style?

File: spec/models/installation_spec.rb [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+      installation = create(
+        :installation,
+        subdomain: "ready-with-different-service"
+      )
+      create(
+        :greenhouse_connection,
+        :connected,
+        :with_namely_field,
+        installation: installation
+      )
```
Comment (seanpdoyle) [inline spec/models/installation_spec.rb]: Could these unassigned calls to FactoryGirl be wrapped up in a helper that describes how they're associated?

File: spec/models/installation_spec.rb [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+        expect(net_suite_connection).to be_a NetSuite::Connection
+        expect(net_suite_connection).to be_persisted
+        expect(net_suite_connection.installation_id).to eq installation.id
+      end
+    end
+  end
+
+  describe "#send_connection_notification" do
+    it "delegates to each users" do
+      users = [build_stubbed(:user), build_stubbed(:user)]
```
Comment (seanpdoyle) [inline spec/models/installation_spec.rb]: We're using this exact line in multiple specs.

What do you think about extracting a `build_users` helper?

If the quantity of users built matters (it doesn't seem to), the helper could accept a count.

Comment (seanpdoyle) [inline spec/models/installation_spec.rb]: Alternatively, we could use a `build_stubbed_list(2, :user)` (if such a thing exists)

File: spec/factories.rb (hunk 6 above)
Comment (jferris) [inline spec/factories.rb]: Yes; these factory changes will perform unnecessary database I/O.

I did this in the interest of trying to minimize changes in the specs, but I'll take a look at removing the `create` calls introduced here.

File: spec/models/user_spec.rb [hunk 14]
```diff
@@ -141,6 +55,65 @@
...
+      expect(Namely::Connection).to have_received(:new).with(
+        access_token: "MY_ACCESS_TOKEN",
+        subdomain: "ellingsonmineral",
+      )
+    end
+  end
+
+  describe "#namely_fields_by_label" do
+    it "returns mappable fields from a Namely connection" do
+      ["single_select", "short_text", "long_text", "number"]
```
Comment (seanpdoyle) [inline spec/models/user_spec.rb]: What is going on with this line?

It looks like an array that isn't assigned, and is inaccessible elsewhere.

File: spec/factories.rb (hunk 6 above)
Comment (jferris) [inline spec/factories.rb]: It introduced some noise, but it's not too bad. I removed the `create` calls from factory definitions in 8dc17a2.

File: app/services/candidate_importer.rb (hunk 8 above)
Comment (jferris) [inline app/services/candidate_importer.rb]: Can you explain the advantage you see to memoizing here? As I see it:

* Memoization prevents mutations from being lost (we aren't mutating)
* Memoization prevents recalculations (associations are cached by ActiveRecord)
* Memoization is slightly more complex
* Memoization is less readable
* Memoization can introduce bugs

In general, I avoid memoization unless I know there's a performance problem I'm addressing.

File: spec/models/installation_spec.rb (hunk 12 above)
Comment (jferris) [inline spec/models/installation_spec.rb]: Yeah, I extracted this from `user_spec.rb`, but it looks pretty weird now that I read it more closely. I'll take a look at it.

Comment (jferris) [inline spec/models/installation_spec.rb]: Addressed in 58b08e4.

File: app/jobs/sync_job.rb (hunk 7 above)
Comment (jferris) [inline app/jobs/sync_job.rb]: Done in ab0d547.

File: spec/models/installation_spec.rb (hunk 9 above)
Comment (jferris) [inline spec/models/installation_spec.rb]: Done in 58b08e470160c5a5ddb21c7d668887c7921835a4.

File: spec/models/authentication_notifier_spec.rb (hunk 10 above)
Comment (jferris) [inline spec/models/authentication_notifier_spec.rb]: I'm not sure it's worthwhile. There are enough dynamic portions in the string that order changes will cause failures, and I think the string version is more readable than a regular expression.

File: spec/models/installation_spec.rb (hunk 11 above)
Comment (jferris) [inline spec/models/installation_spec.rb]: Yeah, I went for local consistency, but it is a little weird to see both in the same diff.

I don't want to rewrite every use of `should`, as that's not what's really going on in this changeset, but I switched all uses affected by these changes to use `is_expected.to` in 74c218d.

File: app/services/candidate_importer.rb (hunk 8 above)
Comment (kenyonj) [inline app/services/candidate_importer.rb]: ok, I wasn't sure what `installation` was doing on the `connection` but if it's not calculating anything or re-pulling something from the db then I agree.

File: spec/models/user_spec.rb (hunk 14 above)
Comment (jferris) [inline spec/models/user_spec.rb]: This was extracted from `attribute_mapper_spec.rb`. As you pointed out, it doesn't do anything. Removed in 4cda597.

File: spec/models/installation_spec.rb (hunk 13 above)
Comment (jferris) [inline spec/models/installation_spec.rb]: I believe this is incidental duplication, rather than duplicated logic. I find the duplicate call to be clearer than the `*_list` versions.

**Comment by jferris** [comment]
@kenyonj @sikachu @seanpdoyle ready for re-review.

**Comment by seanpdoyle** [comment]
LGTM
